src/fhirtypes/common.ts

```diff
@@ -346,14 +368,76 @@ export function applyMixinRules(
         }
         return true;
       });
-      mixedInRules.push(...rules);
+      mixedInRules.push(...(rules as SdRule[]));
     } else {
       logger.error(`Unable to find definition for RuleSet ${mixinName}.`, fshDefinition.sourceInfo);
     }
   });
   fshDefinition.rules = [...mixedInRules, ...fshDefinition.rules];
 }
 
+/**
+ * Adds insert rules onto a Profile, Extension, or Instance
+ * @param fshDefinition - The definition to apply rules on
+ * @param tank - The FSHTank containing the fshDefinition
+ */
+export function applyInsertRules(
+  fshDefinition: Profile | Extension | Instance | FshValueSet | FshCodeSystem | Mapping | RuleSet,
+  tank: FSHTank,
+  seenRuleSets: string[] = []
+): void {
+  const expandedRules: Rule[] = [];
+  fshDefinition.rules.forEach(rule => {
+    if (rule instanceof InsertRule) {
+      const ruleSet = tank.fish(rule.ruleSet, Type.RuleSet) as RuleSet;
+      if (ruleSet) {
+        if (seenRuleSets.includes(ruleSet.name)) {
+          logger.error(
+            `Inserting ${ruleSet.name} will cause a circular dependency, so the rule will be ignored`,
+            rule.sourceInfo
+          );
+        } else {
+          seenRuleSets.push(ruleSet.name);
+          applyInsertRules(ruleSet, tank, seenRuleSets);
+          ruleSet.rules.forEach(ruleSetRule => {
+            // On the import side, a rule that is intended to be a ValueSetConceptComponent can
+            // be imported as a ConceptRule because the syntax is identical. If this is the case,
+            // create a ValueSetConceptComponent that corresponds to the ConceptRule, and use that
+            if (
+              fshDefinition instanceof FshValueSet &&
+              ruleSetRule instanceof ConceptRule &&
+              ruleSetRule.definition == null
```

Comment:
As it is now, if someone mistakenly puts a definition in a code in a `ValueSet`, then this `if` is not satisfied and it will end up being detected as an unsupported rule type in line 424 below -- and generate an error that `ConceptRule` cannot be applied to a `ValueSet`.  I think authors might be confused by this...

I think it would be better remove `ruleSetRule.definition == null` in the `if` statement and then do a special check inside the `if` statement and produce a more appropriate warning: `ValueSet concepts should not include a definition, only system, code, and display are supported.` (or something like that).  And still convert it to a `ValueSetConceptComponentRule` (but without the definition).